### DISCUSSION TITLE
Things to improve upon

### DISCUSSION THREAD

**Discussion by funlennysub** [tokio-rs/toasty#36] [Ideas]
I always been using sea-orm as my go-to ORM for Rust but as every other project it has some cons and pros. Im making this discussion in hopes of helping toasty have less problems I personally deal with in sea-orm.

## Things I like, but can be improved:
- As mentioned in [this discussion](https://github.com/tokio-rs/toasty/discussions/22), migrations as code are a good thing, but the way sea-orm does migrations is too verbose imo, I feel like using toasty's schema language can be useful for migrations, because you don't have to write hundreds of lines simply to create a few tables. In case users need something toasty doesn't yet support (create a postgres function for example), there should probably be a way to have both, schema-defined migration and a code one at the same time in one migration
![image](https://github.com/user-attachments/assets/bffb853a-b38a-4ec0-809c-c34ff80cd2fd)
 
 ## Things I don't really like:
 - The way it handles relations in select statements. Currently, if you want to get all posts made by a certain user, you have a few ways to do that: Find a user you want, then use its field used for relation to find all posts made by them; OR Use a JOIN statement. Both ways have their own drawbacks. If you use the first one, then you create multiple queries which might not be always acceptable, and toasty already does a better job with how you load related tables, then the second method might be better, which brings me to my second point:
 - The way you have to handle a response that doesn't match any of your defined schemas. Again there's 2 ways to do that, a struct that implements `FromQueryResult` or an "unstructured tuple", in case of your own struct, you end up defining some fields multiple times which can lead to problems when table definition changes, as of writing this I don't have a solution that can solve this problem, but I hope people reading this might. For example, in the image below the only difference between, `Query` schema and that struct, is the `query` field, because in the table it is stored as i64 and linked to a table with all names
 ![image](https://github.com/user-attachments/assets/63e0857b-ebfe-4ad7-b587-0178b6b10724)

 
 ## Things sea-orm doesn't have but I needed:
 - A way to do `COPY` statements, for that I have to add another crate just to execute one query which is just pointless. I also have to convert sea-orm column types into `tokio-postgres`', I also have to make sure I didn't make a mistake in `COPY` statement since that crate doesn't know my schema, for that I have my own method that creates that string with all columns from that table

![main code](https://github.com/user-attachments/assets/bbbcf6d6-e05b-43f1-b746-e692cfedb4a9)
![helper methods](https://github.com/user-attachments/assets/e56fdd6a-f266-4f04-aea2-63fd65d9797d)

 
 For now, this seems to be it, but if I remember something else, I will add it here. I think this discussion can be used as a general list of things that can be improved from other ORMs, so feel free to write them down here

**Comment by carllerche** [comment]
Thanks, this is useful. For context, there was a bunch of discussion related to this post here: https://discord.com/channels/500028886025895936/1300487772301168640.

**Comment by bsodmike** [comment]
OT: @funlennysub - Did you write your own driver for Postgres, just wondering and I couldn't find it in the examples.

**Reply by funlennysub** [reply]
I'm not sure what you're talking about
